cpp/src/stream_compaction/stream_compaction.cu

```diff
@@ -0,0 +1,381 @@
+      CUDA_TRY(cudaMemset(null_count, 0, sizeof(gdf_size_type)));
+    }
+
+    bit_mask_t * __restrict__ output_valid =
+      reinterpret_cast<bit_mask_t*>(output_column->valid);
+    bit_mask_t const * __restrict__ input_valid =
+      reinterpret_cast<bit_mask_t*>(input_column->valid);
+
+    scatter<<<grid.num_blocks, block_size>>>(static_cast<T*>(output_column->data),
+                                             output_valid,
+                                             null_count,
+                                             static_cast<T const*>(input_column->data),
+                                             input_valid,
+                                             block_offsets,
+                                             mask_data,
+                                             mask_valid,
+                                             mask_size);
+
+    if (has_valid) {
+      CUDA_TRY(cudaMemcpy(&output_column->null_count, null_count, 
+                          sizeof(gdf_size_type), cudaMemcpyDefault));
+      RMM_FREE(null_count, 0);
+    }
+  }
+};
+
+// Computes the output size of apply_boolean_mask, which is the sum of the 
+// last block's offset and the last block's pass count
+gdf_size_type get_output_size(gdf_size_type *block_counts,
+                              gdf_size_type *block_offsets,
+                              gdf_size_type num_blocks)
+{
+  gdf_size_type last_block_count = 0;
+  cudaMemcpy(&last_block_count, &block_counts[num_blocks - 1], 
+             sizeof(gdf_size_type), cudaMemcpyDefault);
+  gdf_size_type last_block_offset = 0;
+  if (num_blocks > 1)
+    cudaMemcpy(&last_block_offset, &block_offsets[num_blocks - 1], 
+               sizeof(gdf_size_type), cudaMemcpyDefault);
+  return last_block_count + last_block_offset;
+}
+
+/*
+ * Filters a column using a column of boolean values as a mask.
+ *
+ * High Level Algorithm: First, compute a `scatter_map` from the boolean_mask 
+ * that scatters input[i] if boolean_mask[i] is non-null and "true". This is 
+ * simply an exclusive scan of the mask. Second, use the `scatter_map` to
+ * scatter elements from the `input` column into the `output` column.
+ * 
+ * Slightly more complicated for performance reasons: we first compute the 
+ * per-block count of passed elements, then scan that, and perform the
+ * intra-block scan inside the kernel that scatters the output
+ */
+gdf_column apply_boolean_mask(gdf_column const *input,
+                              gdf_column const *boolean_mask) {
+  CUDF_EXPECTS(nullptr != input, "Null input");
+  CUDF_EXPECTS(nullptr != boolean_mask, "Null boolean_mask");
+  CUDF_EXPECTS(input->size == boolean_mask->size, "Column size mismatch");
+  CUDF_EXPECTS(boolean_mask->dtype == GDF_BOOL8, "Mask must be Boolean type");
+  CUDF_EXPECTS(boolean_mask->data != nullptr, "Null boolean_mask data");
+  CUDF_EXPECTS(boolean_mask->valid != nullptr, "Null boolean_mask bitmask");
```

Comment:
I would like to second that we should be able to handle a non-nullable `boolean_mask`

Reply:
Here's a question: do we need to support the case of a `boolean_mask` that is *just* a valid bitmask, and has a null data pointer?  This could allow generating very compact masks, or computing a mask from the AND of the masks of all columns and using that as a filter, etc. It wouldn't be hard to add here.

Reply:
Added support for all 3 types of boolean mask (no valid pointer, no data pointer, or both pointers allocated).